File: c/swatch/swatch.c
/*
  Andrew Pennebaker (andrew.pennebaker@gmail.com)
  Copyright 2007 Andrew Pennebaker

  Credits:

  http://www.krugle.com/files/cvs/cvs.jabberstudio.org/neutron/plugins/time_plugin.py
  For the basic conversion method.

  http://www.acm.uiuc.edu/webmonkeys/book/c_guide/2.15.html
  Thanks for actual working example of C's time and date libraries.
*/

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

static double beats(void) {
  time_t timer;
  struct tm *g;
  int hour, min, sec;
  int utc;
  int bmt;
  double beat;

  timer = time(NULL);

  g = gmtime(&timer);

  hour = g->tm_hour, min = g->tm_min, sec = g->tm_sec;

  utc = hour * 3600 + min * 60 + sec; // Greenwich, England

  bmt = utc + 3600; // Biel, Switzerland

  beat = bmt / 86.4;

  if (beat > 1000) {
    beat -= 1000;
  }

  return beat;
}

static /*@null@*/ char* swatch() {
  char* result;

  result = (char *) malloc(sizeof(char) * 8);

  if (result != NULL) {
    (void) snprintf(result, 8, "@%06.2f", beats());
    return result;
  }
  else {
    printf("Out of memory.\n");
    return NULL;
  }
}

int main() {
  char* s = swatch();

  if (s != NULL) {
    printf("%s\n", s);

    free(s);
  }

  return 0;
}

File: java/swatch/src/main/java/us/yellosoft/swatch/Swatch.java
package us.yellosoft.swatch;

//
// Andrew Pennebaker
// Copyright 2007 Andrew Pennebaker
//
// Credits: http://www.krugle.com/files/cvs/cvs.jabberstudio.org/neutron/plugins/time_plugin.py
//

import java.util.Calendar;

/** Swatch Internet time converter */
public final class Swatch {
  /** utility class */
  private Swatch() {}

  /**
     <p>Get timezone offset in hours</p>
     @param c a calendar
     @return offset
  */
  public static int getTimezone(final Calendar c) {
    return -(c.get(Calendar.ZONE_OFFSET) + c.get(Calendar.DST_OFFSET)) / (60 * 1000) / 60;
  }

  /**
     <p>Get current, raw Swatch beat</p>
     @return unformatted Swatch beat
  */
  public static double beat() {
    Calendar c = Calendar.getInstance();

    int hour = c.get(Calendar.HOUR_OF_DAY) + getTimezone(c);
    int min = c.get(Calendar.MINUTE);
    int sec = c.get(Calendar.SECOND);

    int utc = hour * 3600 + min * 60 + sec; // Greenwich, England

    int bmt = utc + 3600; // Biel, Switzerland

    double beat = bmt / 86.4;

    if (beat > 1000) {
      beat -= 1000;
    }

    return beat;
  }

  /**
     <p>Get current, formatted Swatch time</p>
     @return Swatch time
  */
  public static String swatch() {
    return String.format("@%06.2f", beat());
  }

  /**
     <p>CLI</p>
     @param args CLI args
  */
  public static void main(final String[] args) {
    System.out.println(swatch());
  }
}
